### PR TITLE
Allow blocks to use the role attribute

### LINKED ISSUE #46002: Allow setting role attributes on core blocks
## What problem does this address?

Adding a `role` attribute, (for example `role="button"`) to a core block causes a block validation error.
Allowing the role attribute would increase the flexibility for theme and plugin authors. It would compliment and not be as limited as the HTML element selection option that is in the Advanced panel for supported blocks.

To add a `role` attribute right now you would need to filter the block (or even create a custom block).

## What is your proposed solution?

Similar to the aria label added in https://github.com/WordPress/gutenberg/pull/41744, the `role` should be a block support but not have a UI, because it is an advanced feature.

### PR DESCRIPTION
**Pull request by carolinan**
## What?

This PR:

- Adds a new `role` block support that allows developers to add a `role` attribute in the markup without causing block validation errors.
- Adds the role block support to the group block.

Closes https://github.com/WordPress/gutenberg/issues/46002

## Why?

Allowing the role attribute increases the flexibility for theme and plugin authors. It compliments the HTML element selection option that is in the Advanced panel for supported blocks.

## Testing Instructions

In your WordPress install, locate wp-includes/class-wp-block-supports.php.
On line 184, inside function `get_block_wrapper_attributes,`
add `role` to the array of attributes:
`$attributes_to_merge = array( 'style', 'class', 'id', 'role' );`

Next, login to your WordPress install and create a new post.
Add a group block with a list of links inside.

Open the code editor mode.
Add "role":"navigation" in the block comment.
Add `role="navigation"` to the div:
```
<!-- wp:group {"layout":{"type":"constrained"},"role":"navigation"} -->
<div class="wp-block-group" role="navigation"><!-- wp:list -->
<ul><!-- wp:list-item -->
<li><a href="#">one</a></li>
<!-- /wp:list-item -->

<!-- wp:list-item -->
<li><a href="#">two</a></li>
<!-- /wp:list-item -->

<!-- wp:list-item -->
<li><a href="#">three</a></li>
<!-- /wp:list-item --></ul>
<!-- /wp:list --></div>
<!-- /wp:group -->

```

Save and exit code editor mode.
Confirm that there are no block validation errors.
View source;  confirm that the role on the group block is still `role="document"`

View the post on the front, confirm that the role attribute is present.
This should be tested with and work with themes with and without theme.json.

To test with a dynamic block:
Locate and open the block.json file for a dynamic block, for example the avatar, latest posts or post featured image.
Add `"role":true` to the list of **block supports.**
Build the Gutenberg plugin to apply the change.

Add the block in the block editor.
Open the code editor mode and add the role to the block comment (see example above).
Save and confirm that there are no block validation errors.
View the block on the front and confirm that the role attribute is added correctly.

(Don't forget to revert the change you made to the WordPress file when you are done testing :) )

### CODE REVIEW

**Comment by alexstine** [comment]
@carolinan Just noting that your example should probably read like this.

Old:
```
<div class="wp-block-button" role="button"><a class="wp-block-button__link wp-element-button">test button</a></div>
```

New:
```
<div class="wp-block-button"><a role="button" class="wp-block-button__link wp-element-button">test button</a></div>
```

Adding `role="button"` directly to the div could convert the div itself to a button and we'll have a button link combination.

Thanks.

**Comment by carolinan** [comment]
You are right, I'll try adding an experimental selector.

**Comment by carolinan** [comment]
@gziolo I have not been able to add the role attribute to the link in the button block.  The filter for the save, `'blocks.getSaveContent.extraProps',` adds the props to the block's root element.
Is there any way to do this without creating a new filter?

**Comment by gziolo** [comment]
> @gziolo I have not been able to add the role attribute to the link in the button block. The filter for the save, `'blocks.getSaveContent.extraProps',` adds the props to the block's root element.
> Is there any way to do this without creating a new filter?

As far as I'm aware, blocks that want to avoid injecting the value from the feature/attribute in the root element use `__experimentalSkipSerialization` to mark that feature. I'm not sure whether they can process the full markup and inject that in a different place. I would assume every block author has to handle those non-standard behaviors in the `render_callback` or in `save`.

**Comment by carolinan** [comment]
I apologize for all the back and forward. To summarize:

- I tried using experimental selectors and skipping serialization, but I could not get the block support to work with inner elements, which means I could not make it work with the button block.
- Block attributes were tested instead, but caused problems and failed tests because it altered the role on the block in the editor, which was not desirable.
- Finally the block support was added again, but is only used in the group block.

File: schemas/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+-   Add new block support for the `role` attribute. ([#47868](https://github.com/WordPress/gutenberg/pull/47868/))
+
 -   Add new properties `settings.typography.fluid` and `settings.typography.fontSizes[n].fluidSize` to theme.json to enable fluid typography ([#39529](https://github.com/WordPress/gutenberg/pull/39529)).
```
Comment (carolinan) [inline schemas/CHANGELOG.md:9]: It seems that this changelog has not been maintained? Is it used?

**Comment by fabiankaegy** [comment]
Thank you for working on this! :) 

I have one open question:

Would it make sense to add handling of the block support for Dynamic blocks right away to avoid the issues we are having, for example, extending the `anchor` support to dynamic blocks (https://github.com/WordPress/gutenberg/pull/44771) ?

Yes, this PR only adds the support to the `group` block only. But it does expose the public API of opting into `supports.role` for any block. 

Also since it is a public API it should be added to this reference documentation: https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-supports.md

**Comment by carolinan** [comment]
I can try, but I am not entirely sure what that involves yet :) I'll study the existing files :)

**Comment by carolinan** [comment]
How do I test it with dynamic blocks?
Lets say I want to make sure the featured image is presentational only, and I want to add role="none".
Is this supposed to work in the editor? Is this the goal?
`<!-- wp:post-featured-image {"role":"none" } /-->`

**Comment by fabiankaegy** [comment]
@carolinan Yeah exactly like that :)

**Comment by fabiankaegy** [comment]
And from there if this attribute is present and the block has the block support for it, the `get_block_wrapper_attributes` function will apply all the block supports and the `role="none"` automatically gets added to the markup of the dynamic block on the frontend. 

This is how it was done for the `anchor` support: https://github.com/WordPress/gutenberg/blob/27b1f794bdfc8e3dbe3fe39e1e86c2b58d17225c/lib/block-supports/anchor.php#L30-L67

Which will mean that a patch in WordPress Core is also needed to allow the `role` attribute to be added via a block support: https://github.com/WordPress/wordpress-develop/blob/031cedc06d612c4a74f7b6d3e77d8145f2954a25/src/wp-includes/class-wp-block-supports.php#L184

**Comment by Mamaduka** [comment]
Thanks for working on this, @carolinan!

Maybe we should avoid using a general solution here and use a more direct approach by adding a `role` attribute to the blocks like - Group and Button.

## Why

Currently, there're only two use cases in the core, and not every block can apply a `role` attribute to the root element via `blocks.getSaveContent.extraProps` or `get_block_wrapper_attributes`.

The latter means that consumers have to handle applying the `role` attribute, which defeats the purpose of block type level enhancement.

I think block support-based global enhancements are more useful when they unlock UI, so developers don't have to implement it. Developer-oriented features like this and [`aria-label`](https://github.com/WordPress/gutenberg/pull/41744) can be added as block attributes.

P.S. We should probably avoid introducing server-side support until `role` is supported by `get_block_wrapper_attributes` in the core. Don't want to repeat regression with the `anchor` support.

**Comment by carolinan** [comment]
It does need a core change, but I did not create that ticket since it would be unnecessary if this feature would not be considered.

**Comment by carolinan** [comment]
I could add it as an attribute to the group, as this block has the greatest versatility, where many different roles would be suitable in different situations.
But that still means that other containers like cover would also need to follow up by adding an attribute and not just enabling an existing support.

It would also mean that any custom blocks would need to each make their own implementation instead of using the built-in support?
Finally, it may be just me wanting consistency, and this may be minor? but it would be strange to have the aria-label as a block support and another attribute as a block attribute.

**Comment by Mamaduka** [comment]
> I could add it as an attribute to the group, as this block has the greatest versatility, where many different roles would be suitable in different situations.

I think that would be an excellent first step. Then we can evaluate feedback/usage, introduce the `role` attribute to other container blocks, and eventually enable global support.

We're already taking a similar approach with `allowedBlocks` block attributes until there's a better method to handle block editor settings. See #15682.

It would have been preferable for `aria-label` to be a custom attribute. However, migrating it would require more effort since it is already in the core.

**Pull request closed without merging**